### PR TITLE
JIT: Don't set patchpoints in methods with CORINFO_DEBUG_CODE

### PR DESCRIPTION
**Pull request by AndyAyersMS**
In #88199 the debugger is overriding some jit flags, but has left other flags set that confuse the jit: both `TIER0` and `DEBUG_CODE` end uip set so one part of the jit used logic appropriate for `TIER0` and another part did not.

The JIT ended up creating a patchpoint for OSR in a method with localloc and this combination is not supported by OSR and lead to a crash.

Adjust the runtime so that if optimizations are disabled for a module by the debugger then none of the methods in the module are tiering eligible. This stops the runtime from setting `TIER0` in this case.

Also harden the jit so if the runtime asks for debuggable code, the jit will never generate patchpoints.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@jkotas PTAL
cc @dotnet/jit-contrib

Not obvious to me why disabling R2R also seems to disable optimization, but these settings seem to be linked in VS somehow.

This may not be the complete fix. I am still seeing some methods going through tiering from SPC, so am going to do some more debugging:
```
  13: JIT compiled System.Type:GetTypeFromHandle(System.RuntimeTypeHandle):System.Type [Tier0, IL size=8, code size=40, hash=0x3d27d313]
  14: JIT compiled System.RuntimeType:get_TypeHandle():System.RuntimeTypeHandle:this [Tier0, IL size=7, code size=60, hash=0xc36eeac9]
```
Would be nice to keep this fix fairly minimal so we can back port to 7 if appropriate.

File: src/coreclr/vm/method.cpp [hunk 1]
```diff
@@ -3021,7 +3021,7 @@ bool MethodDesc::DetermineAndSetIsEligibleForTieredCompilation()
         !IsWrapperStub() &&
 
         // Policy - Generating optimized code is not disabled
-        !IsJitOptimizationDisabledForSpecificMethod())
+        !IsJitOptimizationDisabled())
```
Comment (jkotas) [inline src/coreclr/vm/method.cpp:3024]: Is this change necessary?

This code is explicitly factored to minimize the number of flags checked for each method as the comment above explains. We should not get here when the optimizations are disabled globally. 

`DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk` assert above is trying enforce this invariant among other things.

**Comment by jkotas** [comment]
> This may not be the complete fix. I am still seeing some methods going through tiering from SPC, so am going to do some more debugging

Maybe the debugger is disabling the optimizations mid-flight, only after we have JITed a bunch of methods?

**Comment by AndyAyersMS** [comment]
> > This may not be the complete fix. I am still seeing some methods going through tiering from SPC, so am going to do some more debugging
> 
> Maybe the debugger is disabling the optimizations mid-flight, only after we have JITed a bunch of methods?

Could be. I will see if I can understand the dynamics here a bit better.

Just the jit change should be sufficient to fix the bug. But if we only do that, it could lead to some odd behavior at runtime, eg methods getting jitted two (or three, with PGO) times all producing the same debug codegen. But I suspect most users would not notice.

**Comment by AndyAyersMS** [comment]
> > > > This may not be the complete fix. I am still seeing some methods going through tiering from SPC, so am going to do some more debugging
> > > 
> > > 
> > > Maybe the debugger is disabling the optimizations mid-flight, only after we have JITed a bunch of methods?
> > 
> > 
> > Could be. I will see if I can understand the dynamics here a bit better.
> 
> Debug settings do seem to be changing only after determining a whole bunch of methods are tiering-eligible (and a couple of methods have been jitted), but I have yet to pin down how the change happens (it does not seem to go through `SetDebuggerInfoBits`).

The module debug bits seem to get changed out of process via the calls to the debugger initiated by `Debugger::LoadModule`. So we end up in this mixed state where some methods in corelib are tiering-eligible but not optimizable, and others are not tiering-eligible (and not optimizable), and some corelib methods that bypass tiering and are jitted early are optimized.

I am just going to change the jit to fix the specific issue and leave the runtime/debugger behavior alone.

**Review by BruceForstall — APPROVED** [review]

**Review by jkotas — APPROVED** [review]
